build-support/scripts/protobuf.sh

```diff
@@ -104,79 +78,40 @@ function main {
     return 0
 }
 
-# Installs the version of protoc specified by the first argument.
-#
-# Will set 'protoc_bin'
-function protoc_install {
-    local protoc_version="${1:-}"
-    local protoc_os
-
-    if test -z "${protoc_version}"
-    then
-        protoc_version="$(make --no-print-directory print-PROTOC_VERSION)"
-        if test -z "${protoc_version}"
-        then
-            err "ERROR: no protoc-version specified and version could not be discovered"
-            return 1
-        fi
-    fi
-
-    case "$(uname)" in
-        Darwin)
-            protoc_os="osx"
-            ;;
-        Linux)
-            protoc_os="linux"
-            ;;
-        *)
-            err "unexpected OS: $(uname)"
-            return 1
-    esac
-
-    local protoc_zip="protoc-${protoc_version}-${protoc_os}-x86_64.zip"
-    local protoc_url="https://github.com/protocolbuffers/protobuf/releases/download/v${protoc_version}/${protoc_zip}"
-    local protoc_root=".protobuf/protoc-${protoc_os}-${protoc_version}"
-    # This is updated for use outside of the function.
-    protoc_bin="${protoc_root}/bin/protoc"
-
-    if [[ -x "${protoc_bin}" ]]; then
-        status "protocol buffer compiler version already installed: ${protoc_version}"
-        return 0
-    fi
-
-    status_stage "installing protocol buffer compiler version: ${protoc_version}"
-
-    mkdir -p .protobuf/tmp
-    if [[ ! -f .protobuf/tmp/${protoc_zip} ]]; then \
-        ( cd .protobuf/tmp && curl -sSL "${protoc_url}" -o "${protoc_zip}" )
-    fi
-
-    mkdir -p "${protoc_root}"
-    unzip -d "${protoc_root}" ".protobuf/tmp/${protoc_zip}"
-    chmod -R a+Xr "${protoc_root}"
-    chmod +x "${protoc_bin}"
-
-    return 0
-}
-
 function proto_tools_install {
     local protoc_gen_go_version
+    local protoc_gen_go_grpc_version
+    local buf_version
     local mog_version
     local protoc_go_inject_tag_version
 
     protoc_gen_go_version="$(grep github.com/golang/protobuf go.mod | awk '{print $2}')"
+    protoc_gen_go_grpc_version="$(make --no-print-directory print-PROTOC_GEN_GO_GRPC_VERSION)"
     mog_version="$(make --no-print-directory print-MOG_VERSION)"
     protoc_go_inject_tag_version="$(make --no-print-directory print-PROTOC_GO_INJECT_TAG_VERSION)"
+    buf_version="$(make --no-print-directory print-BUF_VERSION)"
 
     # echo "go: ${protoc_gen_go_version}"
     # echo "mog: ${mog_version}"
     # echo "tag: ${protoc_go_inject_tag_version}"
-
+    
+    install_versioned_tool \
+       'buf' \
+        'github.com/bufbuild/buf' \
+        "${buf_version}" \
+        'github.com/bufbuild/buf/cmd/buf'
+        
     install_versioned_tool \
         'protoc-gen-go' \
         'github.com/golang/protobuf' \
         "${protoc_gen_go_version}" \
         'github.com/golang/protobuf/protoc-gen-go'
```

Comment:
I think if we're using the `google.golang.org/grpc` plugin we might want to use the `protoc-gen-go` from the same repository. I don't know if that imposes any implication on us updating our own `go.mod` to point to the `google.golang.org` copy and all that entails though.

Reply:
I tried retargeting everything at google.golang.org/protobuf and it wasn't pretty. There are plenty of breaking changes to account for. So I don't want to do it (at least not in this already giant PR)

However this PR is safe. Both the old grpc generation and new both generate code that pulls in google.golang.org/grpc. The new generator does use `SupportPackageIsVersion7` instead of `SupportPackageIsVersion6` but as the version of the library hasn't changed and it still compiles, it is supposed to mean that compatibility wise everything is fine. There are compile time assertions for those constants in the gRPC library.